fix!: correctly detect if file is outside base path on Windows

fasttime opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [OpenJS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

-   [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Update or create tests
    - If performance-related, include a benchmark
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What is the purpose of this pull request?

Fix logic that detects when a file is outside the base path so it works well with Windows-style paths. Also fixed an edge case with paths that start with ".." but are not external like "..file.js".

#### What changes did you make? (Give an overview)

[Windows paths](https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats) have some unique features not shared by Unix paths. For one, absolute local paths start with a drive letter followed by a colon like "C:". The fact that the drive letter must be the first character in a local path means that a relative path cannot specify a drive. In Node.js, `path.relative()` will just return the absolute path of the second argument when it cannot calculate a relative path across different drives, so for example with `path.relative("C:\\dir", "D:\\file.js")`.

Besides local paths, Windows uses [UNC paths](https://en.wikipedia.org/wiki/Path_(computing)#Universal_Naming_Convention) in the form `\\Server\Volume\dir1\dir2\file` to specify files on a network. Those paths are always absolute.

Both `config-array` and ESLint use relative paths to check if a file is inside a base directory: if the file's relative path from the directory starts with "..", then the file is not in the directory. This check is currently failing when the base directory and the file were located on different drives or servers, because as previously mentioned `path.relative()` doesn't handle those cases.

The fix I've come up with consists in calculating what Node.js calls [namespace-prefixed paths](https://nodejs.org/api/path.html#path_path_tonamespacedpath_path) of the base directory and the file and calling `path.relative()` on those two. Namespaced paths are prefixed with `\\?\` for local paths or with `\\?\UNC` for network paths. When the provided arguments are namespaced paths on different drives, `path.relative()` will return a path in the form `..\..\C:\dir1\dir2\file.js`. While such a path is not valid as it contains a drive letter in middle position, it can be readily recognized as for being outside the base path because of the leading `..` and discarded as external.

#### Related Issues

<!-- include tags like "fixes #123" or "refs #123" -->

- refs https://github.com/eslint/eslint/issues/18575
- preliminary changes in `eslint`: https://github.com/eslint/eslint/compare/update-for-rewrite-pr-59

#### Is there anything you'd like reviewers to focus on?

<!-- markdownlint-disable-file MD004 -->

fasttime commented on eslint.config.js:
```diff
@@ -6,7 +6,7 @@ const eslintPluginJSDoc = eslintConfigESLint.find(
 
 export default [
 	{
-		ignores: ["**/tests/fixtures/", "**/dist/"],
```
Added to fix the CI after the last c8 release.

fasttime commented on packages/config-array/src/config-array.js:
```diff
@@ -351,11 +353,6 @@ function normalizeSync(items, context, extraConfigTypes) {
  */
 function shouldIgnorePath(ignores, filePath, relativeFilePath) {
-	// all files outside of the basePath are ignored
-	if (relativeFilePath.startsWith("..")) {
-		return true;
-	}
```
This check is already performed is `getConfigWithStatus()`.

fasttime commented on packages/config-array/tests/config-array.test.js:
```diff
@@ -10,13 +10,14 @@
 //-----------------------------------------------------------------------------
 
 // calculate base path using import.meta
-const basePath = path.dirname(new URL(import.meta.url).pathname);
```
On Windows, this was giving a path like `/c:/foo/bar/rewrite/packages/config-array/tests`, where the `c:` is considered a top-level directory rather than a drive specifier. This was producing incorrect results with `path.resolve(basePath)` like `C:\c:\foo\bar\rewrite\packages\config-array\tests`.

fasttime commented on packages/config-array/tests/config-array.test.js:
```diff
@@ -2860,6 +3010,30 @@ describe("ConfigArray", () => {
 				);
 			});
 
+			it("should return true when the directory is the parent of the base path", () => {
```
This is to ensure that a relative path like `".."` without a slash is recognized as external.

nzakas commented on packages/config-array/src/config-array.js:
```diff
@@ -386,19 +383,12 @@ function shouldIgnorePath(ignores, filePath, relativeFilePath) {
  * `ignores` only.
  * @param {string} filePath The absolute file path to check.
- * @param {string} basePath The base path for the config.
+ * @param {string} relativeFilePath The relative path of the file to check.
```
Is this relative to the base path still? If so, can you add that to the comment?

fasttime replied:
Good point! I looked at `rollup-plugin-node-polyfills`, the polyfills we use in eslint/eslint, and there is no implementation for `toNamespacedPath` in the [`path`](https://github.com/ionic-team/rollup-plugin-node-polyfills/blob/master/polyfills/path.js) module polyfill, so I added a fallback to return the argument as is when `path.toNamespacedPath` is not defined. This is the same as the implementation on Posix, so basically on every non-Windows system ([source](https://github.com/nodejs/node/blob/92a25abca9dd23a723008d88b1ccb63ff40ec9b7/lib/path.js#L1293-L1300)).

nzakas commented on packages/config-array/src/config-array.js:
```diff
@@ -411,19 +401,12 @@ function pathMatchesIgnores(filePath, basePath, config) {
  * `ignores`.
  * @param {string} filePath The absolute file path to check.
- * @param {string} basePath The base path for the config.
+ * @param {string} relativeFilePath The relative path of the file to check.
```
Same feedback.

fasttime replied:
Updated in 1306ffa88bb514b476caf16a247f86eff55c1f76.

nzakas commented on packages/config-array/src/config-array.js:
```diff
@@ -598,6 +581,13 @@ export class ConfigArray extends Array {
+		// On Windows, `path.relative()` returns an absolute path when given two paths on different drives.
+		// The namespaced base path is useful to make sure that calculated relative paths are always relative.
+		// On Unix, it is identical to the base path.
+		this.namespacedBasePath = path.toNamespacedPath(
```
I was trying to avoid using anything more complicated than `path.relative()` and `path.resolve()`, because I'm not sure if there's a browser polyfill for other methods. Can you verify that `toNamespacedPath()` has an appropriate polyfill?

fasttime replied:
Good point! I looked at `rollup-plugin-node-polyfills`, the polyfills we use in eslint/eslint, and there is no implementation for `toNamespacedPath` in the [`path`](https://github.com/ionic-team/rollup-plugin-node-polyfills/blob/master/polyfills/path.js) module polyfill, so I added a fallback to return the argument as is when `path.toNamespacedPath` is not defined. This is the same as the implementation on Posix, so basically on every non-Windows system ([source](https://github.com/nodejs/node/blob/92a25abca9dd23a723008d88b1ccb63ff40ec9b7/lib/path.js#L1293-L1300)).

fasttime commented on packages/config-array/src/config-array.js:
```diff
@@ -598,6 +586,11 @@ export class ConfigArray extends Array {
+		// On Windows, `path.relative()` returns an absolute path when given two paths on different drives.
+		// The namespaced base path is useful to make sure that calculated relative paths are always relative.
+		// On Unix, it is identical to the base path.
+		this.namespacedBasePath = toNamespacedPath(basePath || process.cwd());
```
`process.cwd()` is always `/` in browser polyfills.

nzakas replied:
I very much don't want to reference `process` in this package. It should not be thinking in terms of cwd because it may not be running in a server-side runtime.

fasttime replied:
I would also prefer not to reference the current directory. The thing is that when `basePath` is an empty string or undefined, `path.relative()` resolves relative paths from the current directory. Do you think we can forbid `""` as a base path and always require that `basePath` is explicitly provided?

nzakas commented on packages/config-array/src/config-array.js:
```diff
@@ -88,8 +88,13 @@ const CONFIG_WITH_STATUS_UNCONFIGURED = Object.freeze({
 
+// Some `node:path` polyfills do not provide an implementation for `toNamespacedPath`.
+// In that case, return the argument unchanged like on POSIX systems.
+const toNamespacedPath = path.toNamespacedPath ?? (arg => arg);
```
I'm confused. If we can just replace this with `arg => arg`, then why should we use `toNamespacedPath` at all?

fasttime replied:
`toNamespacedPath` is only required to handle paths on Windows. Browser polyfills always use Unix-style paths with `/` as a separator and no drive letters, probably because that's how URL paths are written, and in that case it's safe to compare paths without changes. The assumption here is that when `toNamespacedPath` is not defined, we are dealing with Unix paths.

nzakas replied:
Hmm, I'm not sure this is a safe assumption. This package is intended to be used in browsers, in which case we won't know what format the path string is when passed.

Is it possible to inline the `toNamespacedPath` function? Or an approximation?

fasttime replied:
Thanks! I've added a functionally equivalent polyfill for Windows based on [Node.js implementation](https://github.com/nodejs/node/blob/v22.3.0/lib/path.js#L622-L648). Once `toNamespacedPath` is better supported by polyfill libraries, we can simply delete that and switch to using the `node:path` export.

nzakas commented on packages/config-array/src/to-namespaced-path.js:
```diff
@@ -0,0 +1,34 @@
+/**
+ * @fileoverview Polyfill for Node.js `path.toNamespacedPath`
+ * @author Francesco Trotta
```
Just want to double-check: did you write this yourself or is it copied from a polyfill somewhere? (If the latter, we need to include the license.)

fasttime replied:
I wrote it by myself based on the [Node.js implementation](https://github.com/nodejs/node/blob/v22.3.0/lib/path.js#L622-L648). In fact I haven't found any `node:path` polyfill that uses Windows-style paths.

nzakas replied:
Then that means you didn't write it yourself. :) We should include the license from Node.js in this file.

fasttime replied:
Thanks! I've added a link to the Node.js license on GitHub.

nzakas commented on packages/config-array/src/to-namespaced-path.js:
```diff
@@ -0,0 +1,34 @@
+}
+
+// On non-Windows systems, `toNamespacedPath` returns the argument as is.
+export default sep === "\\" ? win32ToNamespacedPath : arg => arg;
```
This doesn't seem like a safe check. Again, because we aren't going to know what format a string is in when it's passed in a browser environment, I don't think we can assume `path.sep` is accurate for the data we'll be getting.

It seems safer to check for the first instance of `\` and `/` in the string? Something like:

```js
const isWinPath = Boolean(filePath.indexOf("/") - file.Path.indexOf("\\"));

fasttime replied:
The input file path could be relative, so there may not be enough indication to determine if it's a Unix or a Windows path. Another complication is that `path.relative` and `path.dirname` have different implementations for Unix and Windows (the path separator `/` vs. `\` is not the only difference). These functions will return different results depending on the platform where Node.js runs even for the same input.

If `ConfigArray` needs to work with both Windows and Unix paths on any platform including browsers, and not depend on `cwd` at all, it would be better to avoid using `node:path` functions to manipulate paths and implement the whole logic internally.

nzakas replied:
> If `ConfigArray` needs to work with both Windows and Unix paths on any platform including browsers, and not depend on `cwd` at all, it would be better to avoid using `node:path` functions to manipulate paths and implement the whole logic internally.

This was actually the intent. Originally, the only reference to `path` I had was `path.relative`, which I intended to replace with an inline function in order to avoid that dependency. My goal going forward was to make sure we weren't starting to use even more of `path` to ensure everything would work in browsers correctly.

Whatever we do, I just don't want to add even deeper dependency on the `path` module. If we need to rethink things a bit, then let's do that.

fasttime replied:
I've started working on generic path handling helpers to handle both Unix and Windows style paths regardless of the current platform. I'm planning to show a prototype in this PR to get some feedback.

nzakas commented:
We may want to consider using the Deno standard path library, which is not runtime or OS specific:
https://jsr.io/@std/path

fasttime commented:
> We may want to consider using the Deno standard path library, which is not runtime or OS specific: https://jsr.io/@std/path

It looks like that library contains different implementations for [windows](https://jsr.io/@std/path/1.0.2/windows) and [posix](https://jsr.io/@std/path/1.0.2/posix), and by default, it selects the implementation for the current platform. That is similar to how Node.js `node:path` works.

It would be useful to find a library that can be installed via npm so we don't need to copy the source files manually.

nzakas commented:
`npx jsr add @std/path` works. :smile:

fasttime commented:
Finally made it work as expected using only the Deno standard library for path handling and without considering the current directory. This fixes the way external files are matched on Windows by including drive letters or UNC server names when resolving relative paths.

An important note: external files are files located outside the base path, but the base path is an optional parameter. If no base path is specified, files will never be considered external. This changes the current behavior where the base path defaults to be the current directory if not specified (in ESLint, the base path is always specified).

nzakas reviewed (dismissed):
LGTM. As I mentioned, I think we should mark this as breaking. 

Would like @mdjermanovic to review before merging.

mdjermanovic commented on .npmrc:
```diff
@@ -1 +1,2 @@
 package-lock = false
+@jsr:registry=https://npm.jsr.io
```
Would end users need to add this to their `.npmrc` files?

mdjermanovic replied:
I've tried using `npm pack`-ed version of config-array from this branch in eslint/eslint, and `npm install` fails with:

```
$ npm i
npm ERR! code E404
npm ERR! 404 Not Found - GET https://registry.npmjs.org/@jsr%2fstd__path - Not found
npm ERR! 404
npm ERR! 404  '@jsr/std__path@^1.0.2' is not in this registry.
npm ERR! 404
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.
```

fasttime replied:
I don't think we can expect users to change their npm settings so that they can download packages from JSR. We could distribute `@std/path` in a separate directory inside `config-array` and import it from there. Or we could use the URL of the tarball from JSR to specify the dependecy in package.json:

```diff
-    "@jsr/std__path": "^1.0.2",
+    "@jsr/std__path": "https://npm.jsr.io/~/11/@jsr/std__path/1.0.2.tgz",
```

Either way, we will lose the magic of the caret `^` to pick newer versions automatically. I checked the JSR docs at https://jsr.io/docs/using-packages but I didn't find any recommendations on how to use their packages with npm apart from registering JSR in .npmrc.

nzakas replied:
Can one of you open an issue on JSR for this? I can't imagine they didn't encounter this before, but before we spin our wheels, it would be nice to verify if we're doing things the right way.

nzakas commented:
I'd say let's just use the URL for now. We can always update the approach if/when they respond to your question.

fasttime commented:
The build is now failing because the generated CommonJS module is trying to `require()` `@jsr/std__path`, which is an ESM-only package. In `packages/config-array/dist/cjs/index.cjs` I can see:

```js
var posixPath = require('@jsr/std__path/posix');
var windowsPath = require('@jsr/std__path/windows');
```

I'm not sure why there was no build error in my previous commit. The error doesn't seem connected to the way `@jsr/std__path` is declared in `package.json`. Maybe we need to bundle `@jsr/std__path` in the CommonJS module so it gets transpiled.

mdjermanovic commented:
> Maybe we need to bundle `@jsr/std__path` in the CommonJS module so it gets transpiled.

If we're going to bundle `@jsr/std__path` in CJS, then we could bundle it in ESM as well, to avoid having a URL depedency (which I think is not a common practice for packages published to npm).

fasttime commented:
> > Maybe we need to bundle `@jsr/std__path` in the CommonJS module so it gets transpiled.
> 
> If we're going to bundle `@jsr/std__path` in CJS, then we could bundle it in ESM as well, to avoid having a URL depedency (which I think is not a common practice for packages published to npm).

Thanks! I tried to bundle `@jsr/std__path` in dist/cjs/index.cjs and dist/esm/index.js, but then `tsc` started to complain because of the `checkJs` option in tsconfig.json:

```console
dist/esm/index.js:316:46 - error TS1313: The body of an 'if' statement cannot be the empty statement.

316       if (lastSlash === i - 1 || dots === 1) ; else if (lastSlash !== i - 1 && dots === 2) {
                                                 ~

dist/esm/index.js:516:15 - error TS2339: Property 'Deno' does not exist on type 'typeof globalThis'.

516       const { Deno } = globalThis;
                  ~~~~

dist/esm/index.js:1148:5 - error TS2353: Object literal may only specify known properties, and 'extended' does not exist in type '{ globstar?: boolean; }'.

1148     extended,
         ~~~~~~~~

dist/esm/index.js:1590:46 - error TS1313: The body of an 'if' statement cannot be the empty statement.

1590       if (lastSlash === i - 1 || dots === 1) ; else if (lastSlash !== i - 1 && dots === 2) {
                                                  ~

dist/esm/index.js:1970:13 - error TS2339: Property 'Deno' does not exist on type 'typeof globalThis'.

1970     const { Deno } = globalThis;
                 ~~~~

dist/esm/index.js:2742:5 - error TS2353: Object literal may only specify known properties, and 'extended' does not exist in type '{ globstar?: boolean; }'.

2742     extended,
         ~~~~~~~~


Found 6 errors in the same file, starting at: dist/esm/index.js:316
```

So I decided to bundle `@jsr/std__path/posix` and `@jsr/std__path/windows` in separate modules that `tsc` would not check. This seems to be working well and results in the following files being generated:

```console
dist/esm/types.d.ts
dist/esm/index.js
dist/esm/types.ts
dist/esm/std__path/posix.js
dist/esm/std__path/windows.js
dist/esm/index.d.ts
dist/cjs/index.d.cts
dist/cjs/types.ts
dist/cjs/index.cjs
dist/cjs/std__path/posix.cjs
dist/cjs/std__path/windows.cjs
```

I will update this PR shortly with the new solution, and hopefully the CI build will pass this time.

fasttime commented on packages/compat/tests/fixup-rules.js:
```diff
@@ -133,7 +134,7 @@ describe("@eslint/backcompat", () => {
 			const code = "var foo = () => 123; function bar() { return 123; }";
 			const messages = linter.verify(code, config, {
-				filename: "test.js",
+				filename: path.resolve("test.js"),
```
When the `filename` parameter is specified, it will be passed as-is to `configArray.getConfig()`, and `getConfig()` expects a full path:

https://github.com/eslint/rewrite/blob/c94ab2dea6410698a4b6814f921b39774b0e56ab/packages/config-array/src/config-array.js#L970

fasttime commented:
Checks are all green now, but I found another issue. In ESLint, when a `Linter` method like `verify()` or `verifyAndFix()` is called without a `filename` parameter, `configArray.getConfig()` gets called with argument `"__placeholder__"` instead of a full path. And if `filename` is a relative path, the relative path will be passed unchanged. Here is the relevant code (https://github.com/eslint/eslint/blob/343f99216096f1db955766870e35d92d5a121448/lib/linter/linter.js#L2010-L2014):

```js
        const filename = options.filename || "__placeholder__.js";

        // Store the config array in order to get plugin envs and rules later.
        internalSlotsMap.get(this).lastConfigArray = configArray;
        const config = configArray.getConfig(filename);
```

Relative paths will no longer work when this PR is merged, and I don't think everybody is using the `Linter` with full paths, because that isn't currently necessary. So we should discuss how to update the `Linter` code as well when we upgrade `config-array` in `ESLint`.

nzakas commented:
To make sure I'm understanding correctly: relative paths will no longer work when passed to `Linter` because they would previously be resolved relative to `process.cwd()`?

fasttime commented:
> To make sure I'm understanding correctly: relative paths will no longer work when passed to `Linter` because they would previously be resolved relative to `process.cwd()`?

Correct. `path.relative()` uses the current directory to resolve relative paths. We are using that logic in multiple places in [packages/config-array/src/config-array.js](https://github.com/eslint/rewrite/blob/main/packages/config-array/src/config-array.js).

nzakas commented:
And again, just for clarity, the `path.relative()` we're using in this PR does *not* use `process.cwd()`?

nzakas commented:
> For a better error message, and also to stay safe in case the `Deno` global is defined for some reason, we could validate path arguments preemptively in `config-array` to make sure that they are absolute paths.

That seems like a good idea...although how would we end up passing an absolute path when `ESLint#lintText()` is called and we use a placeholder? Maybe the placeholder should be `/__placeholder__.js` instead?

fasttime commented:
I've added logic to throw an error when a path is not absolute in a783ba1d19664f18c545bf86c2395f637c73ba98.

fasttime commented:
> how would we end up passing an absolute path when `ESLint#lintText()` is called and we use a placeholder? Maybe the placeholder should be `/__placeholder__.js` instead?

In order to keep the current behavior unchanged we could resolve the `filename` passed to `getConfig()` using `path.resolve()`:

```diff
- const config = configArray.getConfig(filename);
+ const config = configArray.getConfig(path.resolve(filename));
```

Also, we should manually resolve the `cwd` option used to create a `Linter`, because `cwd` is used as a `basePath` for a `ConfigArray`, and we don't want to throw an error when a relative path is passed:

```diff
function normalizeCwd(cwd) {
    if (cwd) {
-       return cwd;
+       return path.resolve(cwd);
    }
    if (typeof process === "object") {
        return process.cwd();
    }
```

I tried it and with those two changes, all `Linter` tests were passing. I also had to update `RuleTester` to make sure that `linter.verify()` was being called with a full path, and fixed a couple of expected paths in the unit tests. And at that point, all tests passed, so I'm confident that this solution will work.

We could also change the way the `Linter` works so it doesn't depend on `process.cwd()` to resolve relative paths, or we could just ban relative paths in the `cwd` option so that `process.cwd()` is not used. Both would be breaking changes I think, but maybe that is where we want to go in ESLint v10?

nzakas commented:
Which file is that first code snippet from?

I don't think we want to create a dependency on `path` inside of `Linter`, as that's a Node-specific API. 

Do we ever pass a relative `cwd` anywhere right now?

fasttime commented:
> Which file is that first code snippet from?

Both snippets are from linter.js in ESLint:

1. https://github.com/eslint/eslint/blob/v9.10.0/lib/linter/linter.js#L2014
2. https://github.com/eslint/eslint/blob/v9.10.0/lib/linter/linter.js#L1200-L1206

> I don't think we want to create a dependency on `path` inside of `Linter`, as that's a Node-specific API.

Actually `Linter` already has [a dependency on `node:path`](https://github.com/eslint/eslint/blob/v9.10.0/lib/linter/linter.js#L14). For `path.resolve()` we could also use an external library like Deno's `@std/path`, or rewrite the logic manually.

> Do we ever pass a relative `cwd` anywhere right now?

On closer look there are only three unit tests where the `cwd` Linter option is a relative path (the string `"cwd"`), so I think we could just fix those tests to use an absolute path like `"/cwd"` instead:

* https://github.com/eslint/eslint/blob/v9.10.0/tests/lib/linter/linter.js#L3869-L3886
* https://github.com/eslint/eslint/blob/v9.10.0/tests/lib/linter/linter.js#L9766-L9786
* https://github.com/eslint/eslint/blob/v9.10.0/tests/lib/linter/linter.js#L9849-L9869

We still need a mechanism to resolve relative file paths in `Linter` for when `config-array` will only accept absolute paths.

nzakas commented:
> Actually `Linter` already has [a dependency on `node:path`](https://github.com/eslint/eslint/blob/v9.10.0/lib/linter/linter.js#L14). For `path.resolve()` we could also use an external library like Deno's `@std/path`, or rewrite the logic manually.

Ah good point. However, the current usage is limited to chopping up filenames rather than resolving. I don't think we can do `@std/path`, because we'd end up needing to bundle like we do here? And that would mean two bundles of `@std/path` in the final `eslint` package.

> We still need a mechanism to resolve relative file paths in `Linter` for when `config-array` will only accept absolute paths.

I'm wondering, could we allow relative paths and just `path.resolve()` it relative to `cwd`? Maybe that would make the problem a bit easier?

fasttime commented:
> However, the current usage is limited to chopping up filenames rather than resolving. I don't think we can do `@std/path`, because we'd end up needing to bundle like we do here? And that would mean two bundles of `@std/path` in the final `eslint` package.

Yes, in order to use `@std/path` in ESLint we would have to add an additional compilation step, so it's better to choose an approach that doesn't need changing the build process.

> I'm wondering, could we allow relative paths and just `path.resolve()` it relative to `cwd`? Maybe that would make the problem a bit easier?

You mean `resolve`ing relative paths in config-array rather that in `Linter`? That should work, as long as a `cwd`/`basePath` is specified (Currently, `basePath` can be omitted in the `ConfigArray` options and it will implicitly default to `process.cwd()`). I can update the PR and test the changes in ESLint.

nzakas commented:
> You mean `resolve`ing relative paths in config-array rather than in `Linter`? That should work, as long as a `cwd`/`basePath` is specified.

Yes, exactly. And sorry, I meant if config array receives a relative path, it should resolve it against the config array's `basePath` (not `cwd`).

> (Currently, `basePath` can be omitted in the `ConfigArray` options and it will implicitly default to `process.cwd()`).

Is that true even after we switched to `@std/path`?

fasttime commented:
> I meant if config array receives a relative path, it should resolve it against the config array's `basePath` (not `cwd`).

Okay, I tried this solution and it seems to work well. In ESLint, only a small change in the `RuleTester` class is required, apart from some tweaks to the unit tests.

Here is what I did:

* In commit 2b97a8162d329700f91e276285f2f98b6b5ff3fa, I added logic to resolve relative paths from the specified `basePath`. If a `basePath` is specified, it must be an absolute path. If `basePath` is not specified, then all other paths passed to `ConfigArray` methods must be absolute paths.
* In commit d9b14789f87f1780bb7d57484d834d5f99f5d6da, I updated the test logic to use relative paths when possible. In the unit tests it is no longer necessary to call `path.resolve(basePath, ...)` or `path.join(basePath, ...)` to create absolute paths when `basePath` is set.
* I prepared a branch in ESLint with the changes necessary to upgrade `config-array` with this PR: https://github.com/eslint/eslint/compare/update-for-rewrite-pr-59. Verified that all tests will pass both with the current version of `config-array` and with the new changes, too.

> > (Currently, `basePath` can be omitted in the `ConfigArray` options and it will implicitly default to `process.cwd()`).
> 
> Is that true even after we switched to `@std/path`?

No. With the changes in this PR, when `basePath` is omitted other paths must be absolute paths, or otherwise an error will be thrown.

nzakas commented:
> * If `basePath` is not specified, then all other paths passed to `ConfigArray` methods must be absolute paths.

Is that necessary? It seems like this should work without a `basePath`. Maybe we could default `basePath` to `/` to allow relative paths to still be passed to the various functions?

fasttime commented:
> Is that necessary? It seems like this should work without a `basePath`. Maybe we could default `basePath` to `/` to allow relative paths to still be passed to the various functions?

It should be fine to use to `/` as a default `basePath` in a `ConfigArray`. Note that this assumes we will be working with Unix-style paths when `basePath` is not specified. I was trying to avoid making that assumption, but after checking the ESLint code, I think this change should be fine, because `Linter` always sets the `basePath` explicitly inside Node.js.

Shall I update the PR to make`/` the default `basePath` ?

nzakas commented:
> Shall I update the PR to make`/` the default `basePath` ?

I think so. @mdjermanovic what do you think?

mdjermanovic commented:
> Shall I update the PR to make`/` the default `basePath` ?

Makes sense to me 👍

fasttime commented:
Thanks for the feedback. I've made `basePath` default to `"/"` in commit 193ece2730eda650859b72274c305a23a613ccef and cleaned up the code used to support a missing base path. Only test changes are required in ESLint in order to use the updated `config-array`: https://github.com/eslint/eslint/compare/update-for-rewrite-pr-59.

fasttime commented on packages/config-array/src/config-array.js:
```diff
@@ -599,6 +626,14 @@ export class ConfigArray extends Array {
 			this.push(configs);
 		}
+
+		// select path-handling implementations depending on the base path
+		this.path = getPathImpl(this.basePath);
+
+		// On Windows, `path.relative()` returns an absolute path when given two paths on different drives.
+		// The namespaced base path is useful to make sure that calculated relative paths are always relative.
+		// On Unix, it is identical to the base path.
+		this.namespacedBasePath = this.path.toNamespacedPath(this.basePath);
```
Should these field be private?

mdjermanovic replied:
I think yes. We could make them public in the future if there's a need for them.

fasttime replied:
Done in 249540676f3a7486d1f7b4cf8637e142fe99bf7a.

mdjermanovic commented on packages/config-array/rollup.std__path-config.js:
```diff
@@ -0,0 +1,28 @@
+export default [
+	{
+		input: "../../node_modules/@jsr/std__path/posix/mod.js",
```
Would it be possible to make this not depend on where the package is installed and also not depend on its internal structure?

fasttime replied:
Done in dd0d81bc55cbe6c7b035af06cab5a7f587b85324.

nzakas commented on packages/config-array/src/config-array.js:
```diff
@@ -551,12 +545,19 @@ const dataCache = new WeakMap();
  * those config objects.
  */
 export class ConfigArray extends Array {
+	/** The namespaced path of the config file directory. */
+	#namespacedBasePath;
+
+	/** Path-handling implementations */
+	#path;
+
```
Can we add types to these properties?

fasttime replied:
Done in 98b9a76bbf1dc74afd8dca5b4548aa8811e7213b.

fasttime replied:
Done in ab61572782b18afdf0b3c3bb1999b641f3d94f2c.

nzakas commented on packages/config-array/src/config-array.js:
```diff
@@ -598,7 +599,7 @@ export class ConfigArray extends Array {
 		 * @type {string}
 		 */
-		this.basePath = basePath;
+		this.basePath = basePath || "/";
```
Just to double check -- should this be nullish or do we also want to check for empty string as an invalid value?

fasttime replied:
I think It's a good idea to treat an empty string as an invalid value, but `RuleTester` is currently setting `basePath` to an empty string when given a relative file path (e.g. `"somefile.js"`).

So if we are going to treat empty strings as invalid we'll need to make a small change in ESLint to ensure that `basePath` is undefined (or `"/"`) in cases where it is currently set to an empty string in `rule-tester.js`, i.e.:

```diff
            if (item.filename) {
-                flatConfigArrayOptions.basePath = path.parse(item.filename).root;
+                flatConfigArrayOptions.basePath = path.parse(item.filename).root || void 0;
            }
```

in https://github.com/eslint/eslint/blob/v9.13.0/lib/rule-tester/rule-tester.js#L655-L657.

Shall we do this or is it better to treat an empty `basePath` as valid and keep the logic in `RuleTester` unchanged?

mdjermanovic replied:
I also think it's a good idea to treat an empty string as an invalid value.

That code was added in RuleTester to support absolute paths. I think it's fine to update it as suggested.

Users that pass `cwd` to `Linter` won't notice the difference as it already treats an empty string the same as undefined (https://github.com/eslint/eslint/blob/1a7cdbfad8d3355132aab3201c7dd3f2e879b19d/lib/linter/linter.js#L1201-L1212).

nzakas replied:
Agreed.

nzakas reviewed (approved):
Overall LGTM. Just left a couple questions for clarification but neither are blockers.

mdjermanovic commented on packages/config-array/src/config-array.js:
```diff
@@ -89,6 +90,9 @@ const CONFIG_WITH_STATUS_UNCONFIGURED = Object.freeze({
 });
 
+// Match two leading dots followed by a slash or the end of input.
+const EXTERNAL_PATH_REGEX = /^\.\.(\/|$)/u;
```
```suggestion
const EXTERNAL_PATH_REGEX = /^\.\.(?:\/|$)/u;
```

Since we're not using the group.

mdjermanovic commented on packages/config-array/src/config-array.js:
```diff
@@ -499,6 +487,44 @@ function assertExtraConfigTypes(extraConfigTypes) {
+/**
+ * Converts a given path to a relative path with all separator characters replaced by forward slashes (`"/"`).
+ * @param {string} fileOrDirPath The unprocessed path to convert.
+ * @param {string} namespacedBasePath The namespaced base path of the directory to which the calulated bath shall be relative.
```
```suggestion
 * @param {string} namespacedBasePath The namespaced base path of the directory to which the calculated path shall be relative.
```

Small typos.

nzakas reviewed (dismissed):
LGTM. Would like @mdjermanovic to verify before merging.

mdjermanovic commented:
> LGTM. Would like @mdjermanovic to verify before merging.

https://github.com/eslint/rewrite/pull/59#discussion_r1806708934 still needs to be addressed by disallowing empty string?

fasttime commented:
> [#59 (comment)](https://github.com/eslint/rewrite/pull/59#discussion_r1806708934) still needs to be addressed by disallowing empty string?

I pushed a commit to disallow empty strings: 4ed3c1c3f6b3665a552aa0f6d762f8daa8bbfd36. Also updated the readme file to reflect the changes: 8c30779592c6402313531e17cfb78a8474e4aea6.

nzakas commented on packages/config-array/src/config-array.js:
```diff
@@ -606,7 +606,7 @@ export class ConfigArray extends Array {
 		 * @type {string}
 		 */
-		this.basePath = basePath || "/";
+		this.basePath = basePath;
```
Can we validate here so we can throw a more specific error, like "basePath must be a non-empty string"?

fasttime replied:
Done in fb712b13ff3a15fbab0117abb858b147ec9e39ff, thanks.

mdjermanovic commented on packages/config-array/tests/config-array.test.js:
```diff
@@ -515,12 +545,12 @@ describe("ConfigArray", () => {
 					name: "from-context",
 				});
 
-				const filename = path.resolve(basePath, "foo.js");
+				const filename = "foo.js";
```
After this change, most of the tests will be with relative paths?

nzakas reviewed (approved):
LGTM pending @mdjermanovic's comment.

mdjermanovic reviewed (approved):
LGTM, thanks! We can revisit testing at some later point.
